src/Transform/ONNX/ConstProp.cpp

```diff
@@ -253,6 +262,130 @@ Value ConstPropWhere(PatternRewriter &rewriter, Value replacingValue,
+  }
+}
+
+std::function<WideNum(WideNum)> divideBy(Type type, int64_t denominator) {
+  return wideZeroDispatchNonBool(type, [denominator](auto wideZero) {
+    using WideCppType = decltype(wideZero);
+    return widenumWrapped<WideCppType, WideCppType>(
+        [denominator](auto x) { return x / denominator; });
+  });
+}
+
+template <typename ReduceOp, typename AxesRange = std::initializer_list<APInt>>
+Value ConstPropReduceAxesRange(PatternRewriter &rewriter, Value replacingValue,
+    Value dataValue, AxesRange axesRange) {
+  ConstPropCounters::count("Reduce", {dataValue});
+  Operation *op = replacingValue.getDefiningOp();
+
+  // Find absoluteAxes, converting any negative axes to non-negative.
+  SmallVector<unsigned, 4> absoluteAxes;
+  ElementsAttr data = getConstValueElements(dataValue);
+  int64_t rank = data.getType().getRank();
+  for (APInt a : axesRange) {
+    int64_t axis = a.getSExtValue();
+    assert(-rank <= axis && axis < rank && "axis out of range");
+    if (axis < 0)
+      axis += rank;
+    assert(std::find(absoluteAxes.begin(), absoluteAxes.end(), axis) ==
+               absoluteAxes.end() &&
+           "duplicate axis");
+    absoluteAxes.push_back(axis);
+  }
+
+  // If axes are empty and !noop_with_empty_axes, reduce over all dimensions.
+  if (absoluteAxes.empty() &&
+      getSIntAttr(op, "noop_with_empty_axes", /*default=*/0) == 0) {
+    for (int64_t axis = 0; axis < rank; ++axis)
+      absoluteAxes.push_back(axis);
+  }
+
+  // Compute the result.
+  ElementsAttr reduced;
+  Type elemType = data.getElementType();
+  if (absoluteAxes.empty()) {
+    reduced = data; // noop
+  } else if (data.empty()) {
+    Attribute identity = getIdentity<ReduceOp>(rewriter, elemType);
+    reduced = DenseElementsAttr::get(replacingValue.getType(), {identity});
+  } else {
+    bool keepdims = getSIntAttr(op, "keepdims", /*default=*/1) != 0;
+    OnnxElementsAttrBuilder elementsBuilder(rewriter.getContext());
+    if constexpr (std::is_same_v<ReduceOp, ONNXReduceMeanOp>) {
+      // sum = ReduceSum(data)
+      ElementsAttr sum = elementsBuilder.reduce(data, absoluteAxes, keepdims,
+          elementwiseBinaryOpCombiner<ONNXAddOp>(elemType));
+      assert(data.size() % sum.size() == 0 &&
+             "ReduceSum reduces tensor size by integer factor");
+      int64_t denominator = data.size() / sum.size();
+      // reduced = sum / denominator
+      reduced = elementsBuilder.transform(
+          sum, elemType, divideBy(elemType, denominator));
+    } else {
+      reduced = elementsBuilder.reduce(data, absoluteAxes, keepdims,
+          elementwiseBinaryOpCombiner<ReduceOp>(elemType));
+    }
+  }
+
+  return createReplacingConstantOp(rewriter, replacingValue, reduced)
+      .getResult();
+}
+
+template <typename ReduceOp>
+Value ConstPropReduce(PatternRewriter &rewriter, Value replacingValue,
+    Value dataValue, Value axesValue) {
+  if (isFromNone(axesValue)) {
+    return ConstPropReduceAxesRange<ReduceOp>(
```

Comment:
Should we call something like initWithIdentity explicitly here? Not sure which way is better.

Reply:
I'm not sure I understand what you mean

fwiw, I'm following NumPy in not relying on an identity for all reduce ops (NumPy doesn't rely on a minimum value as identity for Max or a maximum values as identity for Min, and what would be the identity for Mean? NumPy won't reduce an empty tensor for Max/Min/Mean, whereas for Sum and Prod NumPy reduces an empty tensor to an all-zeros or all-ones result)